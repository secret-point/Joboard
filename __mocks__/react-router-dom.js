// Mock navigate and useParams when needed
// const navigate = jest.fn();
// const useParams = jest.fn();
const useLocation = jest.fn();

module.exports = {
  ...jest.requireActual("react-router-dom"),
  // useNavigate: () => navigate,
  // useParams,
  useLocation,
};
