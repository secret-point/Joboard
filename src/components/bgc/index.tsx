/* eslint-disable react/jsx-wrap-multilines */
import React from "react";
import ReactDOM from "react-dom";

const Index = () => {
  return <div>Index</div>;
};

// export default index;

ReactDOM.render(<Index />, document.getElementById("root"));
