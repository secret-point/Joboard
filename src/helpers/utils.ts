import isEmpty from "lodash/isEmpty";

export const convertPramsToJson = (params: string) => {
  if (!isEmpty(params)) {
    return JSON.parse(
      '{"' +
        decodeURI(params.substring(1))
          .replace(/"/g, '\\"')
          .replace(/&/g, '","')
          .replace(/=/g, '":"') +
        '"}'
    );
  } else {
    return {};
  }
};

export const launchAuthentication = () => {
  let hash = window.location.hash.substr(1);

  const origin = window.location.origin;
  const redirectUrl = `${origin}/#${hash}`;

  const state = window.reduxStore.getState();
  let url = `${
    state.app.appConfig.authenticationURL
  }/?redirectUrl=${encodeURIComponent(redirectUrl)}`;
  window.location.assign(url);
};

export const isJson = (obj: string) => {
  try {
    JSON.parse(obj);
  } catch (e) {
    return false;
  }
  return true;
};
