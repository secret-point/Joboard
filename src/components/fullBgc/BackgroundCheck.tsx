import React, { useEffect } from "react";
import { Button, ButtonVariant } from "@amzn/stencil-react-components/button";
import { Col } from "@amzn/stencil-react-components/layout";
import { H4, Text } from "@amzn/stencil-react-components/text";
import queryString from "query-string";
import { connect } from "react-redux";
import { useLocation } from "react-router-dom";
import { addMetricForPageLoad } from "../../actions/AdobeActions/adobeActions";
import { boundGetCandidateInfo } from "../../actions/CandidateActions/boundCandidateActions";
import { boundGetJobDetail } from "../../actions/JobActions/boundJobDetailActions";
import { boundGetScheduleDetail } from "../../actions/ScheduleActions/boundScheduleActions";
import {
  getPageNameFromPath,
  parseQueryParamsArrayToSingleItem,
  resetIsPageMetricsUpdated
} from "../../helpers/utils";
import { ApplicationState } from "../../reducers/application.reducer";
import { FullBgcState } from "../../reducers/fullBgc.reducer";
import { CandidateState } from "../../reducers/candidate.reducer";
import { JobState } from "../../reducers/job.reducer";
import { ScheduleState } from "../../reducers/schedule.reducer";
import {
  FULL_BGC_STEPS
} from "../../utils/enums/common";
import {
  fullBgcShouldDisplayContinue, getLocale
} from "../../utils/helper";
import { translate as t } from "../../utils/translator";
import { FullBgcStepConfig } from "../../utils/types/common";
import AddressHistory from "./common/AddressHistory";
import BackgroundCheckConsent from "./common/BackgroundCheckConsent";
import BackgroundInformation from "./common/BackgroundInformation";
import BirthHistory from "./common/BirthHistory";
import Documentation from "./common/Documentation";
import InfoStepCard from "../common/InfoStepCard";

interface MapStateToProps {
  job: JobState;
  application: ApplicationState;
  schedule: ScheduleState;
  fullBgc: FullBgcState;
  candidate: CandidateState;
}

type BackgroundCheckMergedProps = MapStateToProps;

export const BackgroundCheck = ( props: BackgroundCheckMergedProps ) => {
  const { job, application, schedule, fullBgc, candidate } = props;
  const { search, pathname } = useLocation();
  const pageName = getPageNameFromPath(pathname);
  const queryParams = parseQueryParamsArrayToSingleItem(queryString.parse(search));
  const { jobId, scheduleId } = queryParams;
  const jobDetail = job.results;
  const applicationData = application.results;
  const stepConfig = fullBgc.stepConfig as FullBgcStepConfig;
  const { scheduleDetail } = schedule.results;
  const { candidateData } = candidate.results;

  useEffect(() => {
    boundGetCandidateInfo();
  }, []);

  useEffect(() => {
    scheduleId && scheduleId !== scheduleDetail?.scheduleId && boundGetScheduleDetail({
      locale: getLocale(),
      scheduleId: scheduleId
    });
  }, [scheduleDetail, scheduleId]);

  // Don't refetch data if id is not changing
  useEffect(() => {
    jobId && jobId !== jobDetail?.jobId && boundGetJobDetail({ jobId: jobId, locale: getLocale() });
  }, [jobDetail, jobId]);

  useEffect(() => {
    // Page will emit page load event once both pros are available but
    // will not emit new event on props change once it has emitted pageload event previously
    scheduleDetail && jobDetail && applicationData && candidateData && addMetricForPageLoad(pageName);
  }, [jobDetail, applicationData, candidateData, scheduleDetail, pageName]);

  useEffect(() => {
    return () => {
      // reset this so as it can emit new pageload event after being unmounted.
      resetIsPageMetricsUpdated(pageName);
    };
  }, [pageName]);

  const handleContinue = () => {
    if (applicationData && scheduleDetail) {
      // TODO: API integration
    }
  };

  return (
    <Col className="bgcContainer" gridGap={15}>
      <Col gridGap={15}>
        <H4>Provide background check information</H4>
        <Text fontSize="T200">
          Please provide authorization and information needed for a background check.
        </Text>
      </Col>

      <InfoStepCard
        title="Background check consent"
        expandedContent={<BackgroundCheckConsent />}
        stepName={FULL_BGC_STEPS.CONSENT}
        infoCardStepStatus={stepConfig[FULL_BGC_STEPS.CONSENT]}
        stepIndex={1}
        maxStepNumber={5}
      />

      <InfoStepCard
        title="Background information"
        expandedContent={<BackgroundInformation />}
        stepName={FULL_BGC_STEPS.BACKGROUND_INFO}
        infoCardStepStatus={stepConfig[FULL_BGC_STEPS.BACKGROUND_INFO]}
        stepIndex={2}
        maxStepNumber={5}
      />

      <InfoStepCard
        title="Address history"
        expandedContent={<AddressHistory />}
        stepName={FULL_BGC_STEPS.ADDRESS_HISTORY}
        infoCardStepStatus={stepConfig[FULL_BGC_STEPS.ADDRESS_HISTORY]}
        stepIndex={3}
        maxStepNumber={5}
      />

      <InfoStepCard
        title="Birth history"
        expandedContent={<BirthHistory />}
        stepName={FULL_BGC_STEPS.BIRTH_HISTORY}
        infoCardStepStatus={stepConfig[FULL_BGC_STEPS.BIRTH_HISTORY]}
        stepIndex={4}
        maxStepNumber={5}
      />

      <InfoStepCard
        title="Documentation"
        expandedContent={<Documentation />}
        stepName={FULL_BGC_STEPS.DOCUMENTATION}
        infoCardStepStatus={stepConfig[FULL_BGC_STEPS.DOCUMENTATION]}
        stepIndex={5}
        maxStepNumber={5}
      />

      <Col padding={{ top: "S300" }}>
        {
          fullBgcShouldDisplayContinue(stepConfig) && (
            <Button
              variant={ButtonVariant.Primary}
              onClick={handleContinue}
            >
              {t("BB-BGC-page-continue-button", "Continue")}
            </Button>
          )}
      </Col>
    </Col>
  );
};

const mapStateToProps = ( state: MapStateToProps ) => {
  return state;
};

export default connect(mapStateToProps)(BackgroundCheck);
