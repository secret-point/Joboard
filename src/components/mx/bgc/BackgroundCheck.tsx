import React, { useEffect } from "react";
import { Col } from "@amzn/stencil-react-components/layout";
import StepHeader from "../../common/StepHeader";
import { JobState } from "../../../reducers/job.reducer";
import { ApplicationState } from "../../../reducers/application.reducer";
import { ScheduleState } from "../../../reducers/schedule.reducer";
import { useLocation } from "react-router-dom";
import {
  getPageNameFromPath,
  parseQueryParamsArrayToSingleItem,
  resetIsPageMetricsUpdated
} from "../../../helpers/utils";
import queryString from "query-string";
import { boundGetJobDetail } from "../../../actions/JobActions/boundJobDetailActions";
import { boundUpdateApplicationDS } from "../../../actions/ApplicationActions/boundApplicationActions";
import {
  bgcMXShouldDisplayContinue,
  checkAndBoundGetApplication,
  createUpdateApplicationRequest,
  getLocale,
  handleUInitiateMXBGCStep
} from "../../../utils/helper";
import { boundGetScheduleDetail } from "../../../actions/ScheduleActions/boundScheduleActions";
import { addMetricForPageLoad } from "../../../actions/AdobeActions/adobeActions";
import { ApplicationStepListDefault } from "../../../utils/constants/common";
import { connect } from "react-redux";
import { H4, Text } from "@amzn/stencil-react-components/text";
import { BGCState } from "../../../reducers/bgc.reducer";
import { BGC_STEPS, PROXY_APPLICATION_STATE, UPDATE_APPLICATION_API_TYPE } from "../../../utils/enums/common";
import { Button, ButtonVariant } from "@amzn/stencil-react-components/button";
import AuthorizationForBGC from "./AuthorizationsForBGC";
import AdditionalBGCInfo from "../../mx/bgc/AdditionalBGCInfo";
import { CandidateState } from "../../../reducers/candidate.reducer";
import { boundGetCandidateInfo } from "../../../actions/CandidateActions/boundCandidateActions";
import { UpdateApplicationRequestDS } from "../../../utils/apiTypes";
import { Application, BgcMXStepConfig } from "../../../utils/types/common";
import { onCompleteTaskHelper } from "../../../actions/WorkflowActions/workflowActions";
import { translate as t } from "../../../utils/translator";
import InfoStepCard from "../../common/InfoStepCard";
import { boundResetBannerMessage } from "../../../actions/UiActions/boundUi";
import { log } from "../../../helpers/log-helper";

interface MapStateToProps {
  job: JobState;
  application: ApplicationState;
  schedule: ScheduleState;
  bgc: BGCState;
  candidate: CandidateState;
}

interface BackgroundCheckProps {

}

type BackgroundCheckMergedProps = MapStateToProps & BackgroundCheckProps;

export const BackgroundCheck = ( props: BackgroundCheckMergedProps ) => {

  const { job, application, schedule, bgc, candidate } = props;
  const { search, pathname } = useLocation();
  const pageName = getPageNameFromPath(pathname);
  const queryParams = parseQueryParamsArrayToSingleItem(queryString.parse(search));
  const { applicationId, jobId, scheduleId } = queryParams;
  const jobDetail = job.results;
  const applicationData = application.results;
  const stepConfig = bgc.stepConfig as BgcMXStepConfig;
  const { scheduleDetail } = schedule.results;
  const { candidateData } = candidate.results;

  useEffect(() => {
    boundGetCandidateInfo();
  }, []);

  useEffect(() => {
    scheduleId && scheduleId!== scheduleDetail?.scheduleId && boundGetScheduleDetail({
      locale: getLocale(),
      scheduleId: scheduleId
    });
  }, [scheduleDetail, scheduleId]);

  // Don't refetch data if id is not changing
  useEffect(() => {
    jobId && jobId !== jobDetail?.jobId && boundGetJobDetail({ jobId: jobId, locale: getLocale() });
  }, [jobDetail, jobId]);

  useEffect(() => {
    checkAndBoundGetApplication(applicationId);
  }, [applicationId]);

  useEffect(() => {
    // Page will emit page load event once both pros are available but
    // will not emit new event on props change once it has emitted pageload event previously
    scheduleDetail && jobDetail && applicationData && candidateData && addMetricForPageLoad(pageName);

  }, [jobDetail, applicationData, candidateData, scheduleDetail, pageName]);

  useEffect(() => {
    return () => {
      // reset this so as it can emit new pageload event after being unmounted.
      resetIsPageMetricsUpdated(pageName);
    };
  }, [pageName]);

  useEffect(() => {
    if (applicationData && candidateData) {
      handleUInitiateMXBGCStep(applicationData, candidateData);
    }
  }, [candidateData, applicationData]);

  const handleContinue = () => {
    boundResetBannerMessage();
    if (applicationData) {
      const { BGC } = UPDATE_APPLICATION_API_TYPE;
      const payload = {
        state: PROXY_APPLICATION_STATE.ADDITIONAL_BACKGROUND_INFO_SAVED
      };
      const request: UpdateApplicationRequestDS = createUpdateApplicationRequest(applicationData, BGC, payload);
      boundUpdateApplicationDS(request, (applicationData: Application) => {
        // For troubleshooting issue: https://issues.amazon.com/issues/HVHBB-1440
        log(`Start to complete bgc workflow step for application ${applicationData.applicationId}, schedule: ${scheduleDetail?.scheduleId}`,
          {
            application: { ...applicationData },
            scheduleDetail: { ...scheduleDetail }
          }
        );
        onCompleteTaskHelper(applicationData);
      });
    }
  };

  return (
    <Col className="bgcContainer" gridGap={15}>
      <StepHeader jobTitle={jobDetail?.jobTitle || ""} step={ApplicationStepListDefault[1]} />
      <Col gridGap={15}>
        <H4>{t("BB-MX-BGC-page-content-title", "Consent to employment screening")}</H4>
        <Text fontSize="T200">
          {t("BB-BGC-page-provide-authorization-and-information-notice", "Please provide authorization and information needed for a background check.")}
        </Text>
      </Col>

      <InfoStepCard
        title={t("BB-BGC-page-authorizations-for-background-check-step-card-title", "Authorizations for Background Check, Medical Check, and Drug Test")}
        expandedContent={<AuthorizationForBGC />}
        stepName={BGC_STEPS.NON_FCRA}
        infoCardStepStatus={stepConfig[BGC_STEPS.NON_FCRA]}
        stepIndex={1}
        maxStepNumber={2}
      />

      <InfoStepCard
        title={t("BB-BGC-page-additional-bgc-step-card-title", "Additional Background Information")}
        expandedContent={<AdditionalBGCInfo />}
        stepName={BGC_STEPS.ADDITIONAL_BGC}
        infoCardStepStatus={stepConfig[BGC_STEPS.ADDITIONAL_BGC]}
        stepIndex={2}
        maxStepNumber={2}
      />
      <Col padding={{ top: "S300" }}>
        {
          bgcMXShouldDisplayContinue(stepConfig) && (
            <Button
              variant={ButtonVariant.Primary}
              onClick={handleContinue}
            >
              {t("BB-BGC-page-continue-button", "Continue")}
            </Button>
          )}
      </Col>
    </Col>
  );
};

const mapStateToProps = ( state: MapStateToProps ) => {
  return state;
};

export default connect(mapStateToProps)(BackgroundCheck);
