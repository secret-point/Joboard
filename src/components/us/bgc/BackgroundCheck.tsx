import React, { useEffect } from "react";
import { Col } from "@amzn/stencil-react-components/layout";
import StepHeader from "../../common/StepHeader";
import { JobState } from "../../../reducers/job.reducer";
import { ApplicationState } from "../../../reducers/application.reducer";
import { ScheduleState } from "../../../reducers/schedule.reducer";
import { useLocation } from "react-router";
import { getPageNameFromPath, parseQueryParamsArrayToSingleItem } from "../../../helpers/utils";
import queryString from "query-string";
import { boundGetJobDetail } from "../../../actions/JobActions/boundJobDetailActions";
import {
    boundGetApplication,
    boundUpdateApplicationDS
} from "../../../actions/ApplicationActions/boundApplicationActions";
import {
    bgcShouldDisplayContinue,
    createUpdateApplicationRequest,
    getLocale,
    handleUInitiateBGCStep,
    routeToAppPageWithPath
} from "../../../utils/helper";
import { boundGetScheduleDetail } from "../../../actions/ScheduleActions/boundScheduleActions";
import { addMetricForPageLoad } from "../../../actions/AdobeActions/adobeActions";
import { ApplicationStepList } from "../../../utils/constants/common";
import { connect } from "react-redux";
import { H4, Text } from "@amzn/stencil-react-components/text";
import { BGCState } from "../../../reducers/bgc.reducer";
import {
    BGC_STEPS,
    INFO_CARD_STEP_STATUS,
    PROXY_APPLICATION_STATE,
    UPDATE_APPLICATION_API_TYPE
} from "../../../utils/enums/common";
import { CommonColors } from "../../../utils/colors";
import { Button, ButtonVariant } from "@amzn/stencil-react-components/button";
import NonFcraDisclosure from "../../common/bgc/NonFcraDisclosure";
import AdditionalBGCInfo from "../../common/bgc/AdditionalBGCInfo";
import { PAGE_ROUTES } from "../../pageRoutes";
import { CandidateState } from "../../../reducers/candidate.reducer";
import { boundGetCandidateInfo } from "../../../actions/CandidateActions/boundCandidateActions";
import { UpdateApplicationRequestDS } from "../../../utils/apiTypes";
import { Application } from "../../../utils/types/common";
import { onCompleteTaskHelper } from "../../../actions/WorkflowActions/workflowActions";
import { translate as t } from "../../../utils/translator";
import InfoStepCard from "../../common/InfoStepCard";

interface MapStateToProps {
    job: JobState,
    application: ApplicationState,
    schedule: ScheduleState,
    bgc: BGCState,
    candidate: CandidateState
}

interface BackgroundCheckProps {

}

type BackgroundCheckMergedProps = MapStateToProps & BackgroundCheckProps;

const BackgroundCheck = ( props: BackgroundCheckMergedProps ) => {

    const { job, application, schedule, bgc, candidate } = props;
    const { search, pathname } = useLocation();
    const pageName = getPageNameFromPath(pathname);
    const queryParams = parseQueryParamsArrayToSingleItem(queryString.parse(search));
    const { applicationId, jobId, scheduleId } = queryParams;
    const jobDetail = job.results;
    const applicationData = application.results;
    const { stepConfig } = bgc;
    const scheduleDetail = schedule.results.scheduleDetail;
    const { candidateData } = candidate.results;
    const { BACKGROUND_CHECK_FCRA } = PAGE_ROUTES;

    useEffect(() => {
        boundGetCandidateInfo();
    },[])

    useEffect(() => {
        scheduleId && boundGetScheduleDetail({
            locale: getLocale(),
            scheduleId: scheduleId
        })
    }, [scheduleId]);

    // Don't refetch data if id is not changing
    useEffect(() => {
        jobId && jobId !== jobDetail?.jobId && boundGetJobDetail({ jobId: jobId, locale: getLocale() })
    }, [jobId]);

    useEffect(() => {
        applicationId && boundGetApplication({ applicationId: applicationId, locale: getLocale() });
    }, [applicationId]);

    useEffect(() => {
        jobDetail && applicationData && scheduleDetail && addMetricForPageLoad(pageName);
    }, [jobDetail, applicationData]);

    useEffect(() => {
        if(applicationData && candidateData) {
            handleUInitiateBGCStep(applicationData, candidateData);
        }
    }, [candidateData, applicationData])

    const handleContinue = () => {
        if(applicationData) {
            const { BGC } = UPDATE_APPLICATION_API_TYPE;
            const payload = {
                state: PROXY_APPLICATION_STATE.ADDITIONAL_BACKGROUND_INFO_SAVED
            }
            const request: UpdateApplicationRequestDS = createUpdateApplicationRequest(applicationData, BGC, payload);
            boundUpdateApplicationDS(request, (applicationData: Application)=>{
                onCompleteTaskHelper(applicationData);
            });
        }
    }

    const renderFCRAExpandedContent = (): React.ReactNode => {
        return (
            <>
                <Col gridGap={15}>
                    <Text
                        fontSize="T200"
                        color={CommonColors.Neutral70}
                    >
                        {t('BB-BGC-page-fcra-subtitle-text','Please answer a few questions using the link below.')}
                    </Text>
                    <Col padding='S300'>
                        <Button
                            variant={ButtonVariant.Primary}
                            onClick={() => routeToAppPageWithPath(BACKGROUND_CHECK_FCRA)}
                        >
                            {stepConfig[BGC_STEPS.FCRA].status === INFO_CARD_STEP_STATUS.ACTIVE ?
                                t("BB-BGC-page-fcra-get-started-btn", "Get Started") :
                                t("BB-BGC-page-fcra-edit-form-btn","Edit Form")
                            }
                        </Button>

                    </Col>
                </Col>
            </>

        )
    }

    return (
        <Col className="bgcContainer" gridGap={15}>
            <StepHeader jobTitle={jobDetail?.jobTitle || ''} step={ApplicationStepList[1]}/>
            <Col gridGap={15}>
                <H4>{t('BB-BGC-page-content-title', 'Background Check')}</H4>
                <Text fontSize="T200">
                    {t('BB-BGC-page-provide-authorization-and-information-notice', 'Please provide authorization and information needed for a background check.')}
                </Text>
            </Col>

            <InfoStepCard
                title={t('BB-BGC-page-fcra-step-card-title', 'Fair Credit Report Act Disclosure')}
                expandedContent={renderFCRAExpandedContent()}
                stepName={BGC_STEPS.FCRA}
                infoCardStepStatus={stepConfig[BGC_STEPS.FCRA]}
                stepIndex={1}
            />

            <InfoStepCard
                title={t('BB-BGC-page-nonfcra-step-card-title', 'Non-Fair Credit Reporting Act Acknowledgments and Authorizations for Background Check')}
                expandedContent={<NonFcraDisclosure/>}
                stepName={BGC_STEPS.NON_FCRA}
                infoCardStepStatus={stepConfig[BGC_STEPS.NON_FCRA]}
                stepIndex={2}
            />

            <InfoStepCard
                title={t('BB-BGC-page-additional-bgc-step-card-title', 'Additional Background Information')}
                expandedContent={<AdditionalBGCInfo/>}
                stepName={BGC_STEPS.ADDITIONAL_BGC}
                infoCardStepStatus={stepConfig[BGC_STEPS.ADDITIONAL_BGC]}
                stepIndex={3}
            />
            <Col padding={{top: 'S300'}}>
                {
                    bgcShouldDisplayContinue(stepConfig) &&
                        <Button
                            variant={ButtonVariant.Primary}
                            onClick={handleContinue}
                        >
                            {t('BB-BGC-page-continue-button', 'Continue')}
                        </Button>
                }
            </Col>
        </Col>
    )
}

const mapStateToProps = ( state: MapStateToProps ) => {
    return state;
};

export default connect(mapStateToProps)(BackgroundCheck);
