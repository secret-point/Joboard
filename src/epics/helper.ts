import { log, LoggerType } from "../helpers/log-helper";
import { ProxyApiError } from "../utils/api/types";

export const epicSwitchMapHelper = async (response: any) => {
  if (!response.data || !!response.errorCode || !!response.error || response.errorMessage) {
    log("[Epic] epicSwitchMapHelper: Starting to raise proxy api error", response, LoggerType.ERROR);
    const { errorCode } = response;
    throw createProxyApiEpicError(errorCode, response.error || response.errorMessage);
  }
  return response;
};

export const createProxyApiEpicError = (errorCode: string, errorMessage?: string): ProxyApiError => {
  return { errorCode, errorMessage };
};
