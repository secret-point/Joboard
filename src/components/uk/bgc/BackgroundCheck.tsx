import React, { Suspense, useEffect } from "react";
import { Col } from "@amzn/stencil-react-components/layout";
import { H3 } from "@amzn/stencil-react-components/text";
import queryString from "query-string";
import { connect } from "react-redux";
import { useLocation } from "react-router-dom";
import { addMetricForPageLoad } from "../../../actions/AdobeActions/adobeActions";
import { boundGetCandidateInfo } from "../../../actions/CandidateActions/boundCandidateActions";
import { boundGetJobDetail } from "../../../actions/JobActions/boundJobDetailActions";
import { boundGetScheduleDetail } from "../../../actions/ScheduleActions/boundScheduleActions";
import {
  getPageNameFromPath,
  parseQueryParamsArrayToSingleItem,
  resetIsPageMetricsUpdated
} from "../../../helpers/utils";
import { ApplicationState } from "../../../reducers/application.reducer";
import { BGCState } from "../../../reducers/bgc.reducer";
import { CandidateState } from "../../../reducers/candidate.reducer";
import { JobState } from "../../../reducers/job.reducer";
import { ScheduleState } from "../../../reducers/schedule.reducer";
import { ApplicationStepList } from "../../../utils/constants/common";
import { checkAndBoundGetApplication, getLocale, initiateScheduleDetailOnPageLoad } from "../../../utils/helper";
import { translate as t } from "../../../utils/translator";
import StepHeader from "../../common/StepHeader";
import AdditionalBGCInfo from "./AdditionalBGCInfo";
import { PAGE_ROUTES } from "../../pageRoutes";

interface MapStateToProps {
  job: JobState;
  application: ApplicationState;
  schedule: ScheduleState;
  bgc: BGCState;
  candidate: CandidateState;
}

interface BackgroundCheckProps {
}

type BackgroundCheckMergedProps = MapStateToProps & BackgroundCheckProps;

export const BackgroundCheck = ( props: BackgroundCheckMergedProps ) => {

  const { job, application, schedule, candidate } = props;
  const { search, pathname } = useLocation();
  const pageName = getPageNameFromPath(pathname);
  const queryParams = parseQueryParamsArrayToSingleItem(queryString.parse(search));
  const { applicationId, jobId, scheduleId } = queryParams;
  const jobDetail = job.results;
  const applicationData = application.results;
  const { scheduleDetail } = schedule.results;
  const { candidateData } = candidate.results;

  useEffect(() => {
    // Refresh and add scheduleId in the url from the jobSelected if it doesn't exist from the query param
    if (!scheduleId && applicationData) {
      initiateScheduleDetailOnPageLoad(applicationData, PAGE_ROUTES.ADDITIONAL_INFORMATION);
    }
  }, [applicationData]);

  useEffect(() => {
    boundGetCandidateInfo();
  }, []);

  useEffect(() => {
    scheduleId && scheduleId!== scheduleDetail?.scheduleId && boundGetScheduleDetail({
      locale: getLocale(),
      scheduleId: scheduleId
    });
  }, [scheduleDetail, scheduleId]);

  // Don't refetch data if id is not changing
  useEffect(() => {
    jobId && jobId !== jobDetail?.jobId && boundGetJobDetail({ jobId: jobId, locale: getLocale() });
  }, [jobDetail, jobId]);

  useEffect(() => {
    checkAndBoundGetApplication(applicationId);
  }, [applicationId]);

  useEffect(() => {
    // Page will emit page load event once both pros are available but
    // will not emit new event on props change once it has emitted pageload event previously
    scheduleDetail && jobDetail && applicationData && candidateData && addMetricForPageLoad(pageName);
  }, [jobDetail, applicationData, candidateData, scheduleDetail, pageName]);

  useEffect(() => {
    return () => {
      // reset this so as it can emit new pageload event after being unmounted.
      resetIsPageMetricsUpdated(pageName);
    };
  }, [pageName]);

  return (
    <Col className="bgcContainer" gridGap={15}>
      <StepHeader jobTitle={jobDetail?.jobTitle || ""} step={ApplicationStepList[1]} />
      <Col gridGap={15}>
        <H3>{t("BB-Kondo-BGC-additional-information-title-text", "Additional information")}</H3>
      </Col>

      <Suspense fallback={<></>}>
        <AdditionalBGCInfo />
      </Suspense>
    </Col>
  );
};

const mapStateToProps = ( state: MapStateToProps ) => {
  return state;
};

export default connect(mapStateToProps)(BackgroundCheck);
