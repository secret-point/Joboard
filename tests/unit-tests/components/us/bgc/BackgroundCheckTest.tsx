import React from "react";
import { shallow } from "enzyme";
import { useLocation } from "react-router-dom";
import { BackgroundCheck } from "../../../../../src/components/us/bgc/BackgroundCheck";
import {
  TEST_APPLICATION_ID,
  TEST_APPLICATION_STATE,
  TEST_BGC_STATE,
  TEST_CANDIDATE,
  TEST_CANDIDATE_STATE,
  TEST_JOB_ID,
  TEST_JOB_STATE,
  TEST_SCHEDULE_ID,
  TEST_SCHEDULE_STATE
} from "../../../../test-utils/test-data";
import * as utilHelper from "../../../../../src/utils/helper";

const shouldPrefillAdditionalBgcInfoSpy = jest.spyOn(utilHelper, "shouldPrefillAdditionalBgcInfo");

describe("BackgroundCheck", () => {
  const mockLocation = {
    pathname: "/bgc",
    search: `?jobId=${TEST_JOB_ID}&applicationId=${TEST_APPLICATION_ID}&scheduleId=${TEST_SCHEDULE_ID}`,
    hash: '',
    state: null
  };
  const mockUseLocation = useLocation as jest.Mock;
  mockUseLocation.mockReturnValue(mockLocation);

  beforeEach(() => {
    shouldPrefillAdditionalBgcInfoSpy.mockReset();
  });

  it("should match snapshot  - if address country is USA", () => {
    shouldPrefillAdditionalBgcInfoSpy.mockReturnValue(true);

    const shallowWrapper = shallow(
      <BackgroundCheck
        candidate={TEST_CANDIDATE_STATE}
        job={TEST_JOB_STATE}
        application={TEST_APPLICATION_STATE}
        schedule={TEST_SCHEDULE_STATE}
        bgc={TEST_BGC_STATE}
      />);

    expect(shallowWrapper).toMatchSnapshot();
  });

  it("should match snapshot  - if address country is Canada", () => {
    shouldPrefillAdditionalBgcInfoSpy.mockReturnValue(false);

    const shallowWrapper = shallow(
      <BackgroundCheck
        candidate={TEST_CANDIDATE_STATE}
        job={TEST_JOB_STATE}
        application={TEST_APPLICATION_STATE}
        schedule={TEST_SCHEDULE_STATE}
        bgc={TEST_BGC_STATE}
      />);

    expect(shallowWrapper).toMatchSnapshot();
  });
});
