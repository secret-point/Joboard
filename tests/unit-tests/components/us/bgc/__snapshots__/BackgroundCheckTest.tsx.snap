// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackgroundCheck should match snapshot 1`] = `
<Col
  className="bgcContainer"
  gridGap={15}
>
  <StepHeader
    jobTitle=""
    step={
      Object {
        "stepNumber": 2,
        "title": "Complete required documents",
        "titleTranslationKey": "BB-ApplicationSteps-complete-required-document-text",
      }
    }
  />
  <Col
    gridGap={15}
  >
    <H4
      color="neutral90"
      font="primary"
      fontSize="T400"
      fontWeight="regular"
      hyphens="none"
      margin={0}
    >
      Background Check
    </H4>
    <Text
      font="primary"
      fontSize="T200"
      hyphens="none"
    >
      Please provide authorization and information needed for a background check.
    </Text>
  </Col>
  <Connect(InfoStepCard)
    expandedContent={
      <React.Fragment>
        <ForwardRef
          gridGap={15}
        >
          <ForwardRef
            color="#56616C"
            font="primary"
            fontSize="T200"
            hyphens="none"
          >
            Please answer a few questions using the link below.
          </ForwardRef>
          <ForwardRef
            padding="S300"
          >
            <ForwardRef
              colorScheme="default"
              iconPosition="leading"
              onClick={[Function]}
              shouldUseResponsiveContent={true}
              size="default"
              type="button"
              variant="primary"
            >
              Get Started
            </ForwardRef>
          </ForwardRef>
        </ForwardRef>
      </React.Fragment>
    }
    infoCardStepStatus={
      Object {
        "editMode": false,
        "status": "ACTIVE",
      }
    }
    stepIndex={1}
    stepName="FCRA"
    title="Fair Credit Report Act Disclosure"
  />
  <Connect(InfoStepCard)
    expandedContent={<UNDEFINED />}
    infoCardStepStatus={
      Object {
        "editMode": false,
        "status": "ACTIVE",
      }
    }
    stepIndex={2}
    stepName="NON_FCRA"
    title="Non-Fair Credit Reporting Act Acknowledgments and Authorizations for Background Check"
  />
  <Connect(InfoStepCard)
    expandedContent={<UNDEFINED />}
    infoCardStepStatus={
      Object {
        "editMode": true,
        "status": "ACTIVE",
      }
    }
    stepIndex={3}
    stepName="ADDITIONAL_BGC"
    title="Additional Background Information"
  />
  <Col
    padding={
      Object {
        "top": "S300",
      }
    }
  />
</Col>
`;
